File: solutions/python/general-algos/arrays/spiral_matrix.py
directions = [
    (0, 1),
    (1, 0),
    (0, -1),
    (-1, 0)
]
RIGHT = 0

def get_next_direction_idx(direction_idx):
    num_directions = len(directions)

    return (direction_idx + 1) % num_directions

def is_in_bounds(matrix, row, col):
    num_rows = len(matrix)
    num_cols = len(matrix[0])

    is_row_in_bounds = row >= 0 and row < num_rows
    is_col_in_bounds = col >= 0 and col < num_cols

    return is_row_in_bounds and is_col_in_bounds

class Solution:
    def move_until_hit_visited_or_oob(self, matrix, start_row, start_col, cur_direction_idx, visited, elements_in_spiral_order):
        row = start_row
        col = start_col

        row_change, col_change = directions[cur_direction_idx]

        while is_in_bounds(matrix, row + row_change, col + col_change) and not (row + row_change, col + col_change) in visited:
            row += row_change
            col += col_change

            elements_in_spiral_order.append(matrix[row][col])
            visited.add((row, col))

        return row, col

    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        elements_in_spiral_order = []
        visited = set()

        cur_row = 0
        cur_col = 0

        cur_direction_idx = RIGHT

        elements_in_spiral_order.append(matrix[cur_row][cur_col])
        visited.add((cur_row, cur_col))

        num_rows = len(matrix)
        num_cols = len(matrix[0])

        num_elements_in_matrix = num_rows * num_cols

        while len(visited) < num_elements_in_matrix:
            new_row, new_col = self.move_until_hit_visited_or_oob(
                matrix,
                cur_row,
                cur_col,
                cur_direction_idx,
                visited,
                elements_in_spiral_order,
            )

            cur_direction_idx = get_next_direction_idx(cur_direction_idx)

            cur_row = new_row
            cur_col = new_col

        return elements_in_spiral_order

File: solutions/javascript/general-algos/arrays/spiralMatrix.js
/**
 * @param {number[][]} matrix
 * @return {number[]}
 */
 const directions = [
    [0, 1],
    [1, 0],
    [0, -1],
    [-1, 0]
];

const RIGHT = 0;

const getPositionString = (row, col) => `${row}, ${col}`;

const getNextDirectionIdx = (directionIdx) => {
    const numDirections = directions.length;
    
    return (directionIdx + 1) % numDirections;
};

const isInBounds = (matrix, row, col) => {
    const numRows = matrix.length;
    const numCols = matrix[0].length;
    
    const isRowInBounds = row >= 0 && row < numRows;
    const isColInBounds = col >= 0 && col < numCols;
    
    return isRowInBounds && isColInBounds;
};

const moveUntilHitVisitedOrOob = (
    matrix,
    startRow,
    startCol,
    curDirectionIdx,
    visited,
    elementsInSpiralOrder,
) => {
    let row = startRow;
    let col = startCol;

    const [rowChange, colChange] = directions[curDirectionIdx];

    while (
        isInBounds(matrix, row + rowChange, col + colChange) &&
        !visited.has(getPositionString(row + rowChange, col + colChange))
    ) {
        row += rowChange;
        col += colChange;

        elementsInSpiralOrder.push(matrix[row][col]);
        visited.add(getPositionString(row, col));
    }

    return [row, col];
};

const spiralOrder = (matrix) => {
    const elementsInSpiralOrder = [];
    const visited = new Set();
    
    let curRow = 0;
    let curCol = 0;
    
    let curDirectionIdx = RIGHT;
    
    elementsInSpiralOrder.push(matrix[curRow][curCol]);
    visited.add(getPositionString(curRow, curCol));
    
    const numRows = matrix.length;
    const numCols = matrix[0].length;
    const numElementsInMatrix = numRows * numCols;
    while (visited.size < numElementsInMatrix) {
        const [newRow, newCol] = moveUntilHitVisitedOrOob(
            matrix,
            curRow,
            curCol,
            curDirectionIdx,
            visited,
            elementsInSpiralOrder,
        );
        
        curDirectionIdx = getNextDirectionIdx(curDirectionIdx);
        
        curRow = newRow;
        curCol = newCol;
    }
    
    return elementsInSpiralOrder;
};